### PR TITLE
feat(cache): file cache cleanup

### LINKED ISSUE #13732: Provide mechanism or documentation for cache cleanup
### What would you like Renovate to be able to do?

It would be nice to offer a method of pruning the cache. On a run with and empty cache directory it populates 32M of space in our instance and after 6 months of hourly runs this grows to 88G.

This is also an issue when using GitLab cache which zips -> uploads and downloads -> unzips the cache on each run, so over time runs begin to take longer and longer.

The current workaround it to completely wipe the cache on a regular basis or when a disk fills up.

Alternatively, if this is not implemented as a feature in code then documentation should be updated provide information on how cache can be pruned by running other commands like `find /renovate/cache -type f -mtime +30 -delete` or if this is not possible and the whole cache should be wiped regularly.

### If you have any ideas on how this should be implemented, please tell us here.

I think it would be preferable to have a config option that can enable cache cleanup (true/false). 
A second option to tune age of cache items may also be desirable (e.g. 30d)

### Is this a feature you are interested in implementing yourself?

No

### PR DESCRIPTION
**Pull request by rarkins**
## Changes

Checks file cache for expired items at the end of a run.

## Context

Closes #13732

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [x] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/util/cache/package/file.ts [hunk 1]
```diff
@@ -68,3 +68,53 @@ export function init(cacheDir: string): void {
...
 }
+
+export async function cleanup(): Promise<void> {
+  logger.debug('Checking file package cache for expired items');
+  try {
+    let totalCount = 0;
+    let deletedCount = 0;
+    for await (const item of cacache.ls.stream(cacheFileName)) {
+      totalCount += 1;
+      const cachedItem = item as unknown as cacache.CacheObject;
```
Comment (rarkins) [inline lib/util/cache/package/file.ts:79]: Not sure why I needed to coerce this type

**Comment by rarkins** [comment]
I'll add tests once gotten positive feedback on the implementation and whether it should be an opt-in flag at first. I think it seems safe enough seeing as it's wrapped in a try/catch

File: lib/util/cache/package/file.ts (hunk 1 above)
Comment (viceice) [reply lib/util/cache/package/file.ts:79]: strange. we should add another try-catch inside the loop and delete invalid cache values

File: lib/util/cache/package/file.ts [hunk 2]
```diff
@@ -68,3 +68,53 @@ export function init(cacheDir: string): void {
...
+      logger.debug(`Deleted ${deletedCount} expired cache entries`);
+    } else {
+      logger.debug('No expired entries need deleting');
+    }
+  } catch (err) {
+    logger.warn({ err }, 'Error cleaning up expired file cache');
+  }
+}
+
+export async function cleanup2(): Promise<void> {
```
Comment (viceice) [inline lib/util/cache/package/file.ts:100]: i would prefer the async loop

**Review by viceice — COMMENTED** [review]
can be very time consuming for very big caches

**Comment by rarkins** [comment]
It's at the end of the run and it needs to be cleaned up *sometime*

File: lib/util/cache/package/file.ts (hunk 1 above)
Comment (rarkins) [reply lib/util/cache/package/file.ts:79]: This is like the "cache index" returned by cacache and if it's not the right structure then I'm not sure there's much we can do

**Comment by rarkins** [comment]
> can be very time consuming for very big caches

I think it needs to be done though. And it cleans at the end of a Renovate run.

Worst case:
- It times out or they kill it manually
- It will have cleaned out X entries by that time
- It will have less to do next time

We're also releasing it in a major so if it somehow makes it impossible to use for some, they can hold off upgrading and raise an issue (or just wipe their cache)

File: lib/util/cache/package/file.ts (hunk 1 above)
Comment (viceice) [reply lib/util/cache/package/file.ts:80]: i mean to delete if JSON parse fails

Comment (rarkins) [reply lib/util/cache/package/file.ts:80]: Done now

**Review by viceice — COMMENTED** [review]
something wrong with your merge. too many changes here

**Review by viceice — APPROVED** [review]

**Comment by rarkins** [comment]
Will add tests

**Pull request merged**